File: Pod/Classes/OAStackViewProxy.m
//
//  OAStackViewProxy.m
//  Pods
//
//  Created by Natan Rolnik on 8/5/15.
//
//

#import "OAStackView.h"
#import "OAStackViewProxy.h"

@interface OAStackViewProxy()

@property (nonatomic, strong) UIView *stackView;
@property (nonatomic, strong) NSArray<UIView *> *initialArrangedSubviews;

@property (nonatomic, strong) UIStackView *nativeStackView;
@property (nonatomic, strong) OAStackView *backwardsCompatibleStackView;

@end

@implementation OAStackViewProxy

+ (BOOL)nativeStackViewAvailable
{
    return NSClassFromString(@"UIStackView") != nil;
}

#pragma mark - Initializers

- (instancetype)init
{
    self = [self initWithArrangedSubviews:nil];
    
    [self commonInit];
    
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    
    [self commonInit];
    
    return self;
}

- (instancetype)initWithCoder:(nonnull NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    
    [self commonInit];
    
    return self;
}

- (instancetype)initWithArrangedSubviews:(NSArray *)arrangedSubviews
{
    self = [super initWithFrame:CGRectZero];
    
    self.initialArrangedSubviews = arrangedSubviews;
    
    [self commonInit];
    
    return self;
}

- (void)commonInit
{
    [self createStackView];
    [self addSubview:self.stackView];
}

- (void)createStackView
{
    if (!self.stackView) {
        if ([OAStackViewProxy nativeStackViewAvailable]) {
            self.nativeStackView = [[UIStackView alloc] initWithArrangedSubviews:self.initialArrangedSubviews];
            self.stackView = self.nativeStackView;
        }
        else {
            self.backwardsCompatibleStackView = [[OAStackView alloc] initWithArrangedSubviews:self.initialArrangedSubviews];
            self.stackView = self.backwardsCompatibleStackView;
        }
    }
}

#pragma mark - Runtime

- (NSMethodSignature*)methodSignatureForSelector:(SEL)selector
{
    NSMethodSignature *signature = [super methodSignatureForSelector:selector];
    
    if (!signature) {
        signature = [self.stackView methodSignatureForSelector:selector];
    }
    
    return signature;
}

- (void)forwardInvocation:(NSInvocation *)anInvocation
{
    if ([self.stackView respondsToSelector:[anInvocation selector]]) {
        [anInvocation invokeWithTarget:self.stackView];
    }
    else {
        [super forwardInvocation:anInvocation];
    }
}

#pragma mark - Setters

- (void)setFrame:(CGRect)frame
{
    [super setFrame:frame];
    
    self.stackView.frame = self.bounds;
}

- (void)setAxis:(UILayoutConstraintAxis)axis
{
    _axis = axis;
    
    self.nativeStackView.axis = axis;
    self.backwardsCompatibleStackView.axis = axis;
}

- (void)setDistribution:(UIStackViewDistribution)distribution
{
    _distribution = distribution;
    
    self.nativeStackView.distribution = distribution;
    self.backwardsCompatibleStackView.distribution = distribution;
}

- (void)setAlignment:(UIStackViewAlignment)alignment
{
    _alignment = alignment;
    
    self.nativeStackView.alignment = alignment;
    self.backwardsCompatibleStackView.alignment = alignment;
}

- (void)setSpacing:(CGFloat)spacing
{
    _spacing = spacing;
    
    self.nativeStackView.spacing = spacing;
    self.backwardsCompatibleStackView.spacing = spacing;
}

- (void)setBaselineRelativeArrangement:(BOOL)baselineRelativeArrangement
{
    _baselineRelativeArrangement = baselineRelativeArrangement;
    
    self.nativeStackView.baselineRelativeArrangement = baselineRelativeArrangement;
}

- (void)setLayoutMarginsRelativeArrangement:(BOOL)layoutMarginsRelativeArrangement
{
    _layoutMarginsRelativeArrangement = layoutMarginsRelativeArrangement;
    
    self.nativeStackView.layoutMarginsRelativeArrangement = layoutMarginsRelativeArrangement;
}

@end

File: Pod/Classes/OAStackViewProxy.swift
//
//  OAStackViewProxy.swift
//  Masilotti.com
//
//  Created by Joe Masilotti on 5/4/16.
//  Copyright © 2016 Masilotti.com. All rights reserved.
//

import UIKit

@objc public class OAStackViewProxy: UIView {
    public init() {
        super.init(frame: CGRectZero)
        if #available(iOS 9, *) {
            installFullScreenView(nativeStackView)
        } else {
            installFullScreenView(backwardsCompatibleStackView)
        }
    }

    public init(arrangedSubviews: [UIView]) {
        super.init(frame: CGRectZero)
        if #available(iOS 9, *) {
            installFullScreenView(nativeStackView)
            arrangedSubviews.forEach({ (view) in nativeStackView.addArrangedSubview(view) })
        } else {
            installFullScreenView(backwardsCompatibleStackView)
            arrangedSubviews.forEach({ (view) in backwardsCompatibleStackView.addArrangedSubview(view) })
        }
    }

    public required init?(coder aDecoder: NSCoder) { fatalError("Unimplemented.") }

    override public var subviews: [UIView] {
        if #available(iOS 9, *) {
            return nativeStackView.subviews
        } else {
            return backwardsCompatibleStackView.subviews
        }
    }

    public var arrangedSubviews: [UIView] {
        if #available(iOS 9, *) {
            return nativeStackView.arrangedSubviews
        } else {
            return backwardsCompatibleStackView.arrangedSubviews
        }
    }

    public func addArrangedSubview(view: UIView) {
        if #available(iOS 9, *) {
            nativeStackView.addArrangedSubview(view)
        } else {
            backwardsCompatibleStackView.addArrangedSubview(view)
        }
    }

    public func removeArrangedSubview(view: UIView) {
        if #available(iOS 9, *) {
            nativeStackView.removeArrangedSubview(view)
        } else {
            backwardsCompatibleStackView.removeArrangedSubview(view)
        }
    }

    public var axis: UILayoutConstraintAxis = .Horizontal {
        didSet {
            if #available(iOS 9, *) {
                nativeStackView.axis = axis
            } else {
                backwardsCompatibleStackView.axis = axis
            }
        }
    }

    public var distribution: OAStackViewDistribution = .Fill {
        didSet {
            if #available(iOS 9, *) {
                nativeStackView.distribution = UIStackViewDistribution(distribution: distribution)
            } else {
                backwardsCompatibleStackView.distribution = distribution
            }
        }
    }

    public var alignment: OAStackViewAlignment = .Fill {
        didSet {
            if #available(iOS 9, *) {
                nativeStackView.alignment = UIStackViewAlignment(alignment: alignment)
            } else {
                backwardsCompatibleStackView.alignment = alignment
            }
        }
    }

    public var spacing: CGFloat = 0.0 {
        didSet {
            if #available(iOS 9, *) {
                nativeStackView.spacing = spacing
            } else {
                backwardsCompatibleStackView.spacing = spacing
            }
        }
    }

    public var baselineRelativeArrangement: Bool = false {
        didSet {
            if #available(iOS 9, *) {
                nativeStackView.baselineRelativeArrangement = baselineRelativeArrangement
            }
        }
    }

    public var layoutMarginsRelativeArrangement: Bool = false {
        didSet {
            if #available(iOS 9, *) {
                nativeStackView.layoutMarginsRelativeArrangement = layoutMarginsRelativeArrangement
            }
        }
    }

    @available(iOS 9.0, *)
    private lazy var nativeStackView: UIStackView = {
        let stackView = UIStackView()
        stackView.translatesAutoresizingMaskIntoConstraints = false
        return stackView
    }()

    private lazy var backwardsCompatibleStackView: OAStackView = {
        let stackView = OAStackView()
        stackView.translatesAutoresizingMaskIntoConstraints = false
        return stackView
    }()
}

@available(iOS 8, *)
public extension OAStackViewProxy {
    public override var layoutMargins: UIEdgeInsets {
        didSet {
            if #available(iOS 9, *) {
                nativeStackView.layoutMargins = layoutMargins
            } else {
                backwardsCompatibleStackView.layoutMargins = layoutMargins
            }
        }
    }
}

@available(iOS 9, *)
private extension UIStackViewAlignment {
    init(alignment: OAStackViewAlignment) {
        switch alignment {
        case .Fill:
            self = .Fill
        case .Leading:
            self = .Leading
        case .FirstBaseline:
            self = .FirstBaseline
        case .Center:
            self = .Center
        case .Trailing:
            self = .Trailing
        case .Baseline:
            self = .FirstBaseline
        }
    }
}

@available(iOS 9, *)
private extension UIStackViewDistribution {
    init(distribution: OAStackViewDistribution) {
        switch distribution {
        case .Fill:
            self = .Fill
        case .FillEqually:
            self = .FillEqually
        case .FillProportionally:
            self = .FillProportionally
        case .EqualSpacing:
            self = .EqualSpacing
        case .EqualCentering:
            self = .EqualCentering
        }
    }
}

private extension UIView {
    func installFullScreenView(view: UIView) {
        addSubview(view)
        view.translatesAutoresizingMaskIntoConstraints = false

        let views = ["view": view]
        addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("H:|[view]|", options: [], metrics: nil, views: views))
        addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("V:|[view]|", options: [], metrics: nil, views: views))
    }
}
